File: windows/tests/ConnectorControl.Core.Tests/TestSupport/MarshalQueue.cs
using System.Collections.Concurrent;

namespace ConnectorControl.Core.Tests.TestSupport;

/// <summary>
/// A stand-in for the WPF dispatcher: callbacks from watcher/timer threads are
/// queued and run only when the test pumps, so the test thread stays the single
/// owner of AppState exactly like the UI thread does in the app.
/// </summary>
public sealed class MarshalQueue
{
    private readonly ConcurrentQueue<Action> queue = new();

    public int Pending => queue.Count;

    public void Post(Action action) => queue.Enqueue(action);

    /// <summary>Runs everything queued so far; returns how many actions ran.</summary>
    public int Pump()
    {
        var ran = 0;
        while (queue.TryDequeue(out var action))
        {
            action();
            ran++;
        }
        return ran;
    }

    /// <summary>Pumps until the condition holds or the timeout passes.</summary>
    public bool PumpUntil(Func<bool> condition, TimeSpan timeout)
    {
        var deadline = DateTime.UtcNow + timeout;
        while (true)
        {
            Pump();
            if (condition())
            {
                return true;
            }
            if (DateTime.UtcNow >= deadline)
            {
                return false;
            }
            Thread.Sleep(50);
        }
    }
}

File: Tests/ConnectorControlStateTests/TestSupport/MarshalQueue.swift
import Foundation
@testable import ConnectorControlState

/// A stand-in for DispatchQueue.main: closures posted from the watchers' queue,
/// the probe's queue or a restart completion are held until the test pumps,
/// so the test thread stays the single owner of AppState exactly like the
/// main actor does in the app.
final class MarshalQueue: @unchecked Sendable {
    private let lock = NSLock()
    private var queue: [MainActorAction] = []

    var pending: Int { lock.withLock { queue.count } }

    func post(_ action: @escaping MainActorAction) {
        lock.withLock { queue.append(action) }
    }

    /// Runs everything queued so far (and anything they queue); returns how many ran.
    @MainActor
    @discardableResult
    func pump() -> Int {
        var ran = 0
        while let action = lock.withLock({ queue.isEmpty ? nil : queue.removeFirst() }) {
            action()
            ran += 1
        }
        return ran
    }

    /// Pumps until the condition holds or the timeout passes.
    @MainActor
    func pumpUntil(_ condition: () -> Bool, timeout: TimeInterval) -> Bool {
        let deadline = Date().addingTimeInterval(timeout)
        while true {
            pump()
            if condition() { return true }
            if Date() >= deadline { return false }
            Thread.sleep(forTimeInterval: 0.05)
        }
    }
}
